src/shared/users.utils.ts

```diff
@@ -0,0 +1,125 @@
+import { DataSourceInstanceSettings } from '@grafana/data';
+import { BackendSrv } from '@grafana/runtime';
+import { queryUntilComplete } from 'core/utils';
+import { QueryResponse } from 'core/types';
+import { User, QueryUsersRequest, QueryUsersResponse } from './types/QueryUsers.types';
+import { QUERY_USERS_MAX_TAKE, QUERY_USERS_REQUEST_PER_SECOND } from './constants/Users.constants';
+
+/**
+ * Represents a utility class for managing and querying user data.
+ */
+export class UsersUtils {
+  /**
+   * A cached promise that resolves to an array of users.
+   * This cache is used to avoid redundant user data fetches.
+   */
+  private static _usersCache: Promise<User[]> | null = null;
+  /**
+   * A cache that stores a promise resolving to a map of user IDs to user names.   
+   */
+  private static _usersMapCache: Promise<Map<string, string>> | null = null;
+
+  
+  /**
+   * Retrieves the cached promise for the list of users.
+   * If the cache is not initialized, it triggers the loading of users.
+   */
+  public get getUsers(): Promise<User[]> {
+    return UsersUtils._usersCache ?? this.loadUsers();
+  }
+
+  constructor(readonly instanceSettings: DataSourceInstanceSettings, readonly backendSrv: BackendSrv) {}
+
+  /**
+   * Retrieves the cached promise for the map of user IDs to full names.
+   * If the cache is not initialized, it generates the map from the user list.
+   */
+  public async getusersMap(): Promise<Map<string, string>> {
+    if(!UsersUtils._usersMapCache) {
+      const users = await this.getUsers;
+      const userMap = new Map<string, string>();
+      users.forEach(user => {
+        const fullName = UsersUtils.getUserFullName(user);
+        userMap.set(user.id, fullName);
+      });
+      UsersUtils._usersMapCache = Promise.resolve(userMap);
```

Comment:
Please apply the pattern [@rajsite suggested](https://github.com/ni/systemlink-grafana-plugins/pull/226#discussion_r2130632778) here.